nemo/core/nemo_model.py

```diff
@@ -0,0 +1,243 @@
+# ! /usr/bin/python
+# -*- coding: utf-8 -*-
+
+# Copyright (c) 2019-, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import random
+import shutil
+import string
+import tarfile
+from abc import abstractmethod
+from os import path
+from typing import Iterable
+
+from .neural_factory import DeploymentFormat, OperationMode
+from .neural_graph import NeuralGraph
+from .neural_modules import ModuleType, NeuralModule
+from nemo import logging
+
+__all__ = ['NeMoModel']
+
+NEMO_TMP_FOLDER = ".nemo_tmp"
+
+
+class NeMoModel(NeuralModule):
+    """Abstract class representing NeMoModel.
+    A NeMoModel is a kind of neural module which contains other neural modules and logic inside.
+    It typically represents a whole neural network and requires only connections with data layer and loss
+    modules for training.
+
+    The same NeMoModel could be used in training, evaluation or inference regimes. It should adjust itself
+    accordingly.
+    """
+
+    def __call__(self, **kwargs):
+        if self._operation_mode == OperationMode.training or self.operation_mode == OperationMode.both:
+            return self.train_graph(**kwargs)
+
+        else:
+            return self.eval_graph(**kwargs)
```

Comment:
How do you handle the case when the training and evaluation graphs have different inputs?

Reply:
in that case train_graph and eval_graph will have different inputs. And because I use **kwargs in both cases that "should work". However, I haven't tested this.

Reply:
Maybe a better way of phrasing this question is: Why do NeMoModels not have predefined input and output ports?

Reply:
because they can depend on the mode